PaX flags are unset during install time

## Description

The grsecurity pax flag checks on the grub binaries are currently failing. Those don't run in CI—yet; see #3909—so we haven't been carefully observing them. 

## Steps to Reproduce

1. Provision local staging: `make staging`.
2. Run testinfra suite: `molecule verify -s libvirt-staging` (substitute `virtualbox-staging` if appropriate)
3. Observe PaX flag errors
4. Re-run staging logic: `molecule converge -s libvirt-staging`
5. Re-run testinfra suite:  `molecule verify -s libvirt-staging` (substitute `virtualbox-staging` if appropriate)
6. Observe all tests pass [0]

## Expected Behavior

The fresh install story configures PaX flags in a persistent manner.

## Actual Behavior

The fresh install story sets PaX flags, but the are overwritten if updates to grub are available.

## Comments

We can resolve this problem by moving the "common" role _before_ the "grsecurity" role; it should be the first in the play. We'll upgrade all packages, then after updates to grub are installed, the PaX flags will be set. 

In #2741 we moved the grsecurity role up to fail faster. That minor convenience isn't worth the risk of misconfiguration. Since #2703, we've been managing PaX flags on grub for _existing_ installs—but we still have an install time window where they can be overwritten. 

[0] Regarding "all tests pass", I actually saw _one_ still fail, but it's unrelated:

<details>
```
    =================================== FAILURES ===================================
    ____ test_securedrop_application_test_journalist_key[ansible://app-staging] ____
    [gw3] linux2 -- Python 2.7.13 /home/conor/.virtualenvs/sd/bin/python2
    
    File = <class 'testinfra.modules.base.GNUFile'>, Sudo = <sudo>
    
        def test_securedrop_application_test_journalist_key(File, Sudo):
            """
            Ensure the SecureDrop Application GPG public key file is present.
            This is a test-only pubkey provided in the repository strictly for testing.
            """
            pubkey_file = File("{}/test_journalist_key.pub".format(
                securedrop_test_vars.securedrop_data))
            # Sudo is only necessary when testing against app hosts, since the
            # permissions are tighter. Let's elevate privileges so we're sure
            # we can read the correct file attributes and test them.
            with Sudo():
                assert pubkey_file.is_file
    >           assert pubkey_file.user == "root"
    E           AssertionError: assert 'www-data' == 'root'
    E             - www-data
    E             + root
    
    ../testinfra/staging/app-code/test_securedrop_app_code.py:61: AssertionError
    === 1 failed, 423 passed, 13 skipped, 3 xfailed, 1 xpassed in 190.45 seconds ===
```
</details>

`molecule verify -s {libvirt,virtualbox}-staging-xenial` fails with the following errors


```
   =================================== FAILURES ===================================
    __________ test_pax_flags[ansible://app-staging-/usr/sbin/grub-probe] __________
    [gw1] linux2 -- Python 2.7.13 /home/kdas/code/sd/bin/python2
    [XPASS(strict)] PaX flags unset at install time, see issue #3916
    _______ test_pax_flags[ansible://app-staging-/usr/sbin/grub-mkdevicemap] _______
    [gw1] linux2 -- Python 2.7.13 /home/kdas/code/sd/bin/python2
    [XPASS(strict)] PaX flags unset at install time, see issue #3916
    _______ test_pax_flags[ansible://app-staging-/usr/bin/grub-script-check] _______
    [gw1] linux2 -- Python 2.7.13 /home/kdas/code/sd/bin/python2
    [XPASS(strict)] PaX flags unset at install time, see issue #3916
    __________ test_pax_flags[ansible://mon-staging-/usr/sbin/grub-probe] __________
    [gw0] linux2 -- Python 2.7.13 /home/kdas/code/sd/bin/python2
    [XPASS(strict)] PaX flags unset at install time, see issue #3916
    _______ test_pax_flags[ansible://mon-staging-/usr/sbin/grub-mkdevicemap] _______
    [gw1] linux2 -- Python 2.7.13 /home/kdas/code/sd/bin/python2
    [XPASS(strict)] PaX flags unset at install time, see issue #3916
    _______ test_pax_flags[ansible://mon-staging-/usr/bin/grub-script-check] _______
    [gw0] linux2 -- Python 2.7.13 /home/kdas/code/sd/bin/python2
    [XPASS(strict)] PaX flags unset at install time, see issue #3916
    === 6 failed, 447 passed, 13 skipped, 2 xfailed, 2 xpassed in 211.42 seconds ===
```

This is while testing #4080

Turns out the base box used for Xenial testing (and therefore presumably any VM, Trusty included) can result in different behavior here. Examples below:

Using a base box dated 2018-08-24 (admittedly rather old):

```
$ vagrant box list | grep ubuntu-16
bento/ubuntu-16.04 (libvirt, 201808.24.0)
bento/ubuntu-16.04 (virtualbox, 201808.24.0)
bento/ubuntu-16.04 (virtualbox, 201812.27.0)
```

All config tests were passing under Xenial:

``` 
======== 446 passed, 13 skipped, 8 xfailed, 2 xpassed in 148.99 seconds ========
```

After upgrading the box base, however, to one dated 

```
$ vagrant box update --box bento/ubuntu-16.04 --provider virtualbox
$ vagrant mutate bento/ubuntu-16.04 libvirt
$ vagrant box list | grep ubuntu-16
bento/ubuntu-16.04 (libvirt, 201808.24.0)
bento/ubuntu-16.04 (virtualbox, 201808.24.0)
bento/ubuntu-16.04 (libvirt, 201812.27.0)
bento/ubuntu-16.04 (virtualbox, 201812.27.0)
```

Then the PaX tests are seen to fail:

```
    =================================== FAILURES ===================================
    __________ test_pax_flags[ansible://app-staging-/usr/sbin/grub-probe] __________
    [gw2] linux2 -- Python 2.7.13 /home/conor/.virtualenvs/sd/bin/python2
    [XPASS(strict)] PaX flags unset at install time, see issue #3916
    _______ test_pax_flags[ansible://app-staging-/usr/sbin/grub-mkdevicemap] _______
    [gw2] linux2 -- Python 2.7.13 /home/conor/.virtualenvs/sd/bin/python2
    [XPASS(strict)] PaX flags unset at install time, see issue #3916
    _______ test_pax_flags[ansible://app-staging-/usr/bin/grub-script-check] _______
    [gw2] linux2 -- Python 2.7.13 /home/conor/.virtualenvs/sd/bin/python2
    [XPASS(strict)] PaX flags unset at install time, see issue #3916
    __________ test_pax_flags[ansible://mon-staging-/usr/sbin/grub-probe] __________
    [gw3] linux2 -- Python 2.7.13 /home/conor/.virtualenvs/sd/bin/python2
    [XPASS(strict)] PaX flags unset at install time, see issue #3916
    _______ test_pax_flags[ansible://mon-staging-/usr/sbin/grub-mkdevicemap] _______
    [gw3] linux2 -- Python 2.7.13 /home/conor/.virtualenvs/sd/bin/python2
    [XPASS(strict)] PaX flags unset at install time, see issue #3916
    _______ test_pax_flags[ansible://mon-staging-/usr/bin/grub-script-check] _______
```

Same as @kushaldas reported above, and in #4080, and most recently in #4099. While a role run order change would indeed resolve in the clean install story, we should consider managing PaX over time, via paxctld or otherwise, to ensure a stable system state over time.

--- a/install_files/ansible-base/group_vars/all/securedrop
+++ b/install_files/ansible-base/group_vars/all/securedrop
@@ -44,4 +44,4 @@ securedrop_pkg_grsec_xenial:
 
 securedrop_pkg_grsec_focal:
   ver: "5.4.97"
-  depends: "linux-image-5.4.97-grsec-securedrop,linux-image-4.14.188-grsec-securedrop,intel-microcode"
+  depends: "linux-image-5.4.97-grsec-securedrop,intel-microcode"

--- a/install_files/ansible-base/roles/grsecurity/tasks/main.yml
+++ b/install_files/ansible-base/roles/grsecurity/tasks/main.yml
@@ -4,6 +4,7 @@
 - include: check_installation.yml
 
 - include: paxctl.yml
+  when: ansible_distribution_release == "xenial"
   tags:
     - paxctl
     - kernel

--- a/install_files/securedrop-grsec-focal/opt/securedrop/paxctld.conf
+++ b/install_files/securedrop-grsec-focal/opt/securedrop/paxctld.conf
@@ -115,3 +115,6 @@
 # /usr/lib/libreoffice/program/soffice.bin	m
 
 /usr/bin/totem			m
+
+# Disable memprotect for Apache, see 4110 for context.
+/usr/sbin/apache2	m

--- a/molecule/testinfra/common/test_grsecurity.py
+++ b/molecule/testinfra/common/test_grsecurity.py
@@ -19,7 +19,6 @@ def test_ssh_motd_disabled(host):
 
 @pytest.mark.parametrize("package", [
     'linux-image-{}-grsec-securedrop',
-    'paxctl',
     'securedrop-grsec',
 ])
 def test_grsecurity_apt_packages(host, package):
@@ -155,48 +154,67 @@ def test_apt_autoremove(host):
     assert "The following packages will be REMOVED" not in c.stdout
 
 
-@pytest.mark.xfail(reason="PaX flags unset at install time, see issue #3916")
-@pytest.mark.parametrize("binary", [
-    "/usr/sbin/grub-probe",
-    "/usr/sbin/grub-mkdevicemap",
-    "/usr/bin/grub-script-check",
-])
-def test_pax_flags(host, binary):
+def test_paxctl(host):
     """
-    Ensure PaX flags are set correctly on critical Grub binaries.
-    These flags are maintained as part of a post-install kernel hook
-    in the `securedrop-grsec` metapackage. If they aren't set correctly,
-    the machine may fail to boot into a new kernel.
+    As of Focal, paxctl is not used, and shouldn't be installed.
     """
-
-    f = host.file("/etc/kernel/postinst.d/paxctl-grub")
-    assert f.is_file
-    assert f.contains("^paxctl -zCE {}".format(binary))
-
-    c = host.run("paxctl -v {}".format(binary))
-    assert c.rc == 0
-
-    assert "- PaX flags: --------E--- [{}]".format(binary) in c.stdout
-    assert "EMUTRAMP is enabled" in c.stdout
-    # Tracking regressions; previous versions of the Ansible config set
-    # the "p" and "m" flags.
-    assert "PAGEEXEC is disabled" not in c.stdout
-    assert "MPROTECT is disabled" not in c.stdout
-
-
-def test_paxctld(host):
-    """
-    Ensures that paxctld is configured and running. Only relevant
-    for Focal hosts.
-    """
+    p = host.package("paxctl")
     if host.system_info.codename == "xenial":
+        assert p.is_installed
+    else:
+        assert not p.is_installed
+
+
+def test_paxctld_xenial(host):
+    """
+    Xenial-specific paxctld config checks.
+    Ensures paxctld is running and enabled, and relevant
+    exemptions are present in the config file.
+    """
+    if host.system_info.codename != "xenial":
         return True
+    hostname = host.ansible.get_variables()["inventory_hostname"]
+    # Under Xenial, apache2 pax flags managed by securedrop-app-code.
+    if "app" not in hostname:
+        return True
+
     assert host.package("paxctld").is_installed
-    assert host.file("/etc/paxctld.conf").is_file
-    assert host.file("/opt/securedrop/paxctld.conf").is_file
+    f = host.file("/etc/paxctld.conf")
+    assert f.is_file
+    assert f.contains("^/usr/sbin/apache2\tm")
+
     s = host.service("paxctld")
     assert s.is_enabled
     assert s.is_running
+
+
+def test_paxctld_focal(host):
+    """
+    Focal-specific paxctld config checks.
+    Ensures paxctld is running and enabled, and relevant
+    exemptions are present in the config file.
+    """
+    if host.system_info.codename != "focal":
+        return True
+
+    assert host.package("paxctld").is_installed
+    f = host.file("/etc/paxctld.conf")
+    assert f.is_file
+
+    s = host.service("paxctld")
+    assert s.is_enabled
+    assert s.is_running
+
+    # The securedrop-grsec metapackage will copy the config
+    # out of /opt/ to ensure the file is always clobbered on changes.
+    assert host.file("/opt/securedrop/paxctld.conf").is_file
+
+    hostname = host.ansible.get_variables()["inventory_hostname"]
+    # Under Focal, apache2 pax flags managed by securedrop-grsec metapackage.
+    # Both hosts, app & mon, should have the same exemptions. Check precedence
+    # between install-local-packages & apt-test repo for securedrop-grsec.
+    if "app" in hostname:
+        assert f.contains("^/usr/sbin/apache2\tm")
 
 
 @pytest.mark.parametrize('kernel_opts', [

